primitives/keystore/src/lib.rs

```diff
@@ -0,0 +1,385 @@
+	/// an error otherwise.
+	async fn sign_with(
+		&self,
+		id: KeyTypeId,
+		key: &CryptoTypePublicPair,
+		msg: &[u8],
+	) -> Result<Vec<u8>, Error>;
+
+	/// Sign with any key
+	///
+	/// Given a list of public keys, find the first supported key and
+	/// sign the provided message with that key.
+	///
+	/// Returns a tuple of the used key and the SCALE encoded signature.
+	async fn sign_with_any(
+		&self,
+		id: KeyTypeId,
+		keys: Vec<CryptoTypePublicPair>,
+		msg: &[u8]
+	) -> Result<(CryptoTypePublicPair, Vec<u8>), Error> {
+		if keys.len() == 1 {
+			return self.sign_with(id, &keys[0], msg).await.map(|s| (keys[0].clone(), s));
+		} else {
+			for k in self.supported_keys(id, keys).await? {
+				if let Ok(sign) = self.sign_with(id, &k, msg).await {
+					return Ok((k, sign));
+				}
+			}
+		}
+		Err(Error::KeyNotSupported(id))
+	}
+
+	/// Sign with all keys
+	///
+	/// Provided a list of public keys, sign a message with
+	/// each key given that the key is supported.
+	///
+	/// Returns a list of `Result`s each representing the SCALE encoded
+	/// signature of each key or a Error for non-supported keys.
+	async fn sign_with_all(
+		&self,
+		id: KeyTypeId,
+		keys: Vec<CryptoTypePublicPair>,
+		msg: &[u8],
+	) -> Result<Vec<Result<Vec<u8>, Error>>, ()> {
+		let futs = keys.iter()
+			.map(|k| self.sign_with(id, k, msg));
+
+		Ok(join_all(futs).await)
+	}
+
+	/// Generate VRF signature for given transcript data.
+	///
+	/// Receives KeyTypeId and Public key to be able to map
+	/// them to a private key that exists in the keystore which
+	/// is, in turn, used for signing the provided transcript.
+	///
+	/// Returns a result containing the signature data.
+	/// Namely, VRFOutput and VRFProof which are returned
+	/// inside the `VRFSignature` container struct.
+	///
+	/// This function will return an error in the cases where
+	/// the public key and key type provided do not match a private
+	/// key in the keystore. Or, in the context of remote signing
+	/// an error could be a network one.
+	#[cfg(feature = "std")]
+	async fn sr25519_vrf_sign(
+		&self,
+		key_type: KeyTypeId,
+		public: &sr25519::Public,
+		transcript_data: VRFTranscriptData,
+	) -> Result<VRFSignature, Error>;
+}
+
+/// [deprecated] Sync version of the CryptoStore
+///
+/// Some parts of Substrate still rely on a sync version of the `CryptoStore`.
+/// To make the transition easier this auto trait wraps any async `CryptoStore` and
+/// exposes a `sync` interface using `block_on`. Usage of this is deprecated and it
+/// will be removed as soon as the internal usage has transitioned successfully.
+/// If you are starting out building something new **do not use this**,
+/// instead, use [`CryptoStore`].
+pub trait SyncCryptoStore: CryptoStore + Send + Sync {
+	/// Returns all sr25519 public keys for the given key type.
+	fn sr25519_public_keys(&self, id: KeyTypeId) -> Vec<sr25519::Public> {
+		block_on(CryptoStore::sr25519_public_keys(self, id))
+	}
```

Comment:
Hmm.. is it good to have this being a trait? That would mean that `API`s can require _this_ to be implemented, no? From our usage however, it feels like we are using it more like a const fn-struct (is that its name?). Isn't it preferable to do that instead (untested code)?


```suggestion
pub struct SyncCryptoStore;

impl<S: CryptoStore + Send + Sync> SyncCryptoStore {
	/// Returns all sr25519 public keys for the given key type.
	fn sr25519_public_keys(store: &S, id: KeyTypeId) -> Vec<sr25519::Public> {
		block_on(storesr25519_public_keys(id))
	}
	// ...
```

That way, your API can't require you passed a `SyncCryptoStore`, but would have to depend on the helper directly.

Reply:
`SyncCryptoStore` used to be a struct but @bkchr and @mxinden preferred this being a trait instead.

Reply:
That was a bit different though. It was an _instantiated_ struct, yes – but this isn't instantiated. It would not require any additional memory entry or wrapping as you'd pass the type to every function call – it is just a bundle of these functions around a type.

I only care like 4/10 about this, but it seems like a not so clean API without any (to me) obvious benefits but a few drawbacks.

Reply:
> That would mean that APIs can require this to be implemented, no?

Correct. Someone could implement `SyncCryptoStore` for `Foo` in case `Foo` already implements `CryptoStore`. Can you extend on how that is problematic @gnunicorn? Are you worried people override the default method implementations?

> but it seems like a not so clean API without any (to me) obvious benefits but a few drawbacks.

I see the idea of an extension trait a well known pattern, e.g. see the [FutureExt](https://docs.rs/futures/0.3.5/futures/future/trait.FutureExt.html) trait. What are the other drawbacks of this approach that you have in mind @gnunicorn?

Reply:
well, we agreed that this is a temporary thing that shouldn't exist for long, isn't even meant as a helper but _only_ because some stuff we currently have is not implementing them properly. However, it being a `trait` means someone might now does

```rust
impl<S: SyncCryptoStore> A {
    fn open(self, store: &S) {
        store.sync_function()
    }
}
```

but I think the main idea is that we remove it as soon as we don't have usage anymore internally. And I don't see the benefit over having them then do:
```
impl<S: CryptoStore> A {
    fn open(self, store:&S) {
        SyncCryptoStore::sync_function(store);
    }
}
```

> I see the idea of an extension trait a well known pattern, e.g. see the FutureExt trait. 

I am not saying it is bad pattern (by no means), but I feel in this case, where we _do not_ want to support this behaviour long term, the other is cleaner.

But you are right, every `SyncCryptoStore` also has to implement `CryptoStore`, so you can't implement in any way that would harm us down the line. As said, I think the other cleaner but do not block this PR because of that. Go ahead as you see fit.

Reply:
I appreciate the detailed explanation @gnunicorn.

> But you are right, every SyncCryptoStore also has to implement CryptoStore, so you can't implement in any way that would harm us down the line.

With that in mind I don't find the following problematic, as one could replace `SyncCryptoStore` with `CryptoStore` at any point in time.

```rust
impl<S: SyncCryptoStore> A {
    fn open(self, store: &S) {
        store.sync_function()
    }
}
```

> As said, I think the other cleaner but I do not want to block this PR because of that. Go ahead as you see fit.

I am fine either way as well. I have come across the extension trait pattern more often than I have come across the const-fn-struct pattern, not sure that should get a lot of weight in this discussion though.